src/signal.rs

```diff
@@ -14,35 +13,30 @@ pub enum SignalTo {
 /// Signals from OS/user.
 #[cfg(unix)]
 pub fn signals() -> impl Stream<Item = SignalTo> {
-    use tokio_signal::unix::{Signal, SIGHUP, SIGINT, SIGQUIT, SIGTERM};
-
-    let sigint = Signal::new(SIGINT).flatten_stream();
-    let sigterm = Signal::new(SIGTERM).flatten_stream();
-    let sigquit = Signal::new(SIGQUIT).flatten_stream();
-    let sighup = Signal::new(SIGHUP).flatten_stream();
-
-    let signals = sigint.select(sigterm.select(sigquit.select(sighup)));
-
-    signals
-        .map(|signal| match signal {
-            SIGHUP => SignalTo::Reload,
-            SIGINT | SIGTERM => SignalTo::Shutdown,
-            SIGQUIT => SignalTo::Quit,
-            _ => unreachable!(),
-        })
-        .compat()
-        .into_stream()
-        .map(|result| result.expect("Neither stream errors"))
+    use futures::future::{select_all, BoxFuture};
+    use tokio::signal::unix::{signal, SignalKind};
+
+    let mut sigint = signal(SignalKind::interrupt()).expect("Signal handlers should not panic.");
+    let mut sigterm = signal(SignalKind::terminate()).expect("Signal handlers should not panic.");
+    let mut sigquit = signal(SignalKind::quit()).expect("Signal handlers should not panic.");
+    let mut sighup = signal(SignalKind::hangup()).expect("Signal handlers should not panic.");
+
+    let set: Vec<BoxFuture<SignalTo>> = vec![
+        Box::pin(async move { sigint.recv().map(|_| SignalTo::Shutdown).await }),
+        Box::pin(async move { sigterm.recv().map(|_| SignalTo::Shutdown).await }),
+        Box::pin(async move { sigquit.recv().map(|_| SignalTo::Quit).await }),
+        Box::pin(async move { sighup.recv().map(|_| SignalTo::Reload).await }),
+    ];
+
+    let selection = select_all(set.into_iter()).map(|(val, _, _)| val);
+
+    selection.into_stream()
```

Comment:
It looks like we might have changed the behavior here a bit. Before, we had a combined continuous stream of these signals. Now it seems like we have a future of only the first to be received, which is then converted to a single-element stream.

Reply:
Thank you for noticing this. :) I fixed it (I think?) and learnt some great new tricks while doing it.